csrc/id_model/id_model.cpp

```diff
@@ -837,4 +845,275 @@ ValGraph IdModel::buildIntersection(
   return intersection;
 }
 
+namespace {
+
+// Check if there's an equivalent expression as iel_expr that uses
+// maybe_promoted_inputs. This is used to avoid redundantly replaying
+// expressions.
+// NOTE: This is currently overly conservative and some
+// opportunities for reuse are lost, althought it doesn't affect
+// the correctness of the analysis.
+Expr* findMatchingExpr(
+    const ExprGroup& iel_expr,
+    const ValGraph& iel_graph,
+    const std::vector<IterDomain*>& maybe_promoted_inputs) {
+  // Grab all eligible uses of the promoted inputs
+  ExprGroups maybe_promoted_input_uses;
+  // Note that any eligible matching expr should be a use of all
+  // inputs in maybe_promoted_input_uses, no matter it's promoted or
+  // not. So it isn't necessary to look at all of
+  // maybe_promoted_input_uses but just need to grab one.
+  for (auto inp_id : maybe_promoted_inputs) {
+    // inp_id may have been just replayed, in which case it should
+    // not exist in the IEL graph. It should be just ignored as it
+    // should not have any use yet.
+    if (!iel_graph.hasGroup(inp_id)) {
+      continue;
+    }
+    const auto& inp_iel_group = iel_graph.toGroup(inp_id);
+    maybe_promoted_input_uses = iel_graph.getUses(inp_iel_group);
+    break;
+  }
+
+  if (maybe_promoted_input_uses.empty()) {
+    return nullptr;
+  }
+
+  // Look for exprs that have inputs that are mapped in the IEL
+  // graph with the (promoted) inputs of iel_expr.
+  for (const ExprGroup& maybe_promoted_input_use_group :
+       maybe_promoted_input_uses) {
+    NVF_ERROR(!maybe_promoted_input_use_group->empty());
+    // maybe_promoted_inputs may include non-promoted inputs as
+    // well, so maybe_promoted_input_uses may include the original
+    // iel_expr itself. Since there must at least be a promoted input,
+    // iel_expr itself should not be an expr group we are looking for.
+    if (iel_expr == maybe_promoted_input_use_group) {
+      continue;
+    }
+    Expr* maybe_promoted_input_use = maybe_promoted_input_use_group->front();
+    if (!iel_expr->front()->sameOp(maybe_promoted_input_use)) {
+      continue;
+    }
+    // Check if all inputs are mapped
+    NVF_ERROR(
+        maybe_promoted_inputs.size() ==
+        maybe_promoted_input_use->inputs().size());
+    bool all_inputs_match = true;
+    for (const auto inp_i : c10::irange(maybe_promoted_inputs.size())) {
+      // Here, new promoted ids are not added to iel_graph, so
+      // once promoted, this should not return true anymore. Also,
+      // strictAreMapped doesn't work as promoted domains are not
+      // in the graph
```

Comment:
If we have new promoted id, then should we just `return nullptr`?

Reply:
This should no longer matter as that case is filtered out at the beginning of the function. I changed `permissiveAreMapped` to `strictAreMapped`.